rustls/src/verify.rs

```diff
@@ -319,29 +324,22 @@ type CertChainAndRoots<'a, 'b> = (
 );
 
 fn prepare<'a, 'b>(
+    end_entity: &'a Certificate,
+    intermediates: &'a [Certificate],
     roots: &'b RootCertStore,
-    presented_certs: &'a [Certificate],
 ) -> Result<CertChainAndRoots<'a, 'b>, TLSError> {
-    if presented_certs.is_empty() {
-        return Err(TLSError::NoCertificatesPresented);
-    }
-
     // EE cert must appear first.
-    let cert = webpki::EndEntityCert::from(&presented_certs[0].0).map_err(TLSError::WebPKIError)?;
```

Comment:
This is the only use of the `slice[]` indexing operator that I can see, and it seems like it could be replaced by

```
let ee = presented_certs.first().ok_or(TLSError::WebPKIError)?.0;
```

Are there other cases I'm missing? It may be good to change the API for clarity reasons, but it doesn't seem that the API change is needed in order to avoid the indexing operator.

Reply:
It is true there are other ways to avoid the slicing. However, I think this is the best way, as it resolves several issues with the API. I have updated the commit message and the PR description with more details about the improvements to the API, and I've improved the doc comments in the code.